dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/oprovider/DhisOauthAuthenticationProvider.java

```diff
@@ -0,0 +1,155 @@
+import org.hisp.dhis.security.oauth2.DefaultClientDetailsUserDetailsService;
+import org.hisp.dhis.user.UserService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.crypto.password.NoOpPasswordEncoder;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Henning Håkonsen
+ */
+@Slf4j
+@Component
+public class DhisOauthAuthenticationProvider extends DaoAuthenticationProvider
+{
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private SecurityService securityService;
+
+    @Autowired
+    DefaultClientDetailsUserDetailsService defaultClientDetailsUserDetailsService;
+
+    @Autowired
+    public DhisOauthAuthenticationProvider( @Qualifier( "defaultClientDetailsUserDetailsService" ) DefaultClientDetailsUserDetailsService detailsService )
+    {
+//        super();
+        setUserDetailsService( detailsService );
+        setPasswordEncoder( NoOpPasswordEncoder.getInstance() );
+        this.defaultClientDetailsUserDetailsService = detailsService;
+    }
+
+    @Override
+    public Authentication authenticate( Authentication auth )
+        throws AuthenticationException
+    {
+        log.info( String.format( "OldOauthAuthenticationProvider Login attempt: %s", auth.getName() ) );
+
+        String username = auth.getName();
+
+        UserDetails userCredentials = getUserDetailsService().loadUserByUsername( username );
+
+        if ( userCredentials == null )
+        {
+            throw new BadCredentialsException( "Invalid username or password" );
+        }
+
+        // Initialize all required properties of user credentials since these will become detached
+
+//        userCredentials.getAllAuthorities();
+//
+//        // -------------------------------------------------------------------------
+//        // Check two-factor authentication
+//        // -------------------------------------------------------------------------
+//
+//        if ( userCredentials.isTwoFA() )
+//        {
+//            TwoFactorWebAuthenticationDetails authDetails =
+//                (TwoFactorWebAuthenticationDetails) auth.getDetails();
+//
+//            // -------------------------------------------------------------------------
+//            // Check whether account is locked due to multiple failed login attempts
+//            // -------------------------------------------------------------------------
+//
+//            if ( authDetails == null )
```

Comment:
Does two-factor still work?

If yes, remove commented code?

Reply:
Yes, two factor works but only via form login. This provider is for oauth2 access token login